cardano-crypto-class/src/Cardano/Crypto/Libsodium/Memory/Internal.hs

```diff
@@ -130,3 +183,89 @@ sodiumFree ptr size = do
     errno <- getErrno
     ioException $ errnoToIOError "c_sodium_munlock" errno Nothing Nothing
   c_sodium_free ptr
+
+zeroMem :: MonadST m => Ptr a -> CSize -> m ()
+zeroMem ptr size = unsafeIOToMonadST . void $ c_memset (castPtr ptr) 0 size
+
+copyMem :: MonadST m => Ptr a -> Ptr a -> CSize -> m ()
+copyMem dst src size = unsafeIOToMonadST . void $ c_memcpy (castPtr dst) (castPtr src) size
+
+allocaBytes :: Int -> (Ptr a -> ST s b) -> ST s b
+allocaBytes size f =
+  unsafeIOToST $ Foreign.allocaBytes size (unsafeSTToIO . f)
+
+useByteStringAsCStringLen :: ByteString -> (CStringLen -> ST s a) -> ST s a
+useByteStringAsCStringLen bs f =
+  allocaBytes (BS.length bs + 1) $ \buf -> do
+    len <- unsafeIOToST $ BS.unsafeUseAsCStringLen bs $ \(ptr, len) ->
+      len <$ copyMem buf ptr (fromIntegral len)
+    f (buf, len)
```

Comment:
Ah, I thought we had discussed this - the intention here is for this function to be a drop-in replacement for `Data.ByteString.useAsCStringLen`, which is implemented in terms of `useAsCString`, and that one does add a `NUL` terminator.

It's still wrong though, because while we do allocate an extra byte, we fail to actually set it to `NUL`. In any case, this is where that `+ 1` comes from; while we don't actually need anything to be zero-terminated AFAIK, I considered it safer to remain consistent with `useAsCStringLen`, just in case anything depends on it after all.

If you think this won't be a problem, then I'd be OK with either of your versions, provided we document this behavior, otherwise I suggest we fix the `NUL` termination and keep the `+ 1`.

Reply:
I've kept the +1 and fixed the NUL termination.

NUL termination isn't needed in situations where this function is used correctly (i.e., respecting the reported length), but it is unfortunately a reality that functions are not always used correctly, and adding an extra NUL is a cheap and pretty inconsequential way of getting an extra defense layer here (a.k.a. "in-depth security"). The memory and performance overhead are not big enough to worry about IMO, and for correct usages, the extra memory at the end makes absolutely no difference whatsoever, because the NUL byte lives outside the reported length. Of course it would be nicer if we could somehow prevent incorrect usages statically, but I don't think that's feasible at this level (though it is at higher levels, where we have type-level key sizes that we can check).

The function is not currently used, but the `DirectSerialise` API (which will be in another PR) will use it for non-mlocked data.

I think I will rename it anyway, not just to make it look less like `ByteString.useAsCStringLen`, but also because it isn't really "use as", it's "copy to" ("use as" would suggest that you get a pointer to the actual bytestring memory, which you don't).

Reply:
> it is unfortunately a reality that functions are not always used correctly

This function is intended for usage in `cardano-crypo-class` package, so we control how it is used. I'd expect null termination to be used *only* with functions that *require* null termination. I personally don't agree with mantra: let's add a byte here and there just in case someone uses a function incorrectly. Considering that null termination makes no sense for serialization and especially of binary data I don't understand why we even need it. I wish this function was added in a PR that it is actually being used.

That being said, if you wanna keep it because you believe it will be useful later, fine. But let's name it something really obvious, like: `useByteStringAsCStringLenWithNullTermination`. This way in places where it is being used we can ask a question: do we rely here on null termination?

I keep forgetting to mention it: we need to make sure to add **all** of the new functionality to the changelog. Because this function is exported and is considered as part of the public API it should be added as well. Adding some tests would also be very nice, considering that someone might use it and we don't even know if it works as expected.

It is important to understand that we are not writing a memory managing library like `bytestring`. If that was the case then I would say let's have two functions that do both: with null and without null termination, and also document and test them very well. But we are not writing such library, at least not yet, so adding proactively functions that might never be used only adds mental overhead: implementing, review, discussions, testing, documenting, etc. All of these cycles can be spent elsewhere on stuff that actually matters.

Reply:
> This function is intended for usage in cardano-crypo-class package, so we control how it is used.

This, I think, is the core part that I don't agree with - IMO, any function that is exported from a module should be held to the same standards as a "library function", but this here is not the place to argue that.

I'll see if I can move this function to the PR where it is actually used, which should address the "never used" issue, and then hopefully we can get this PR here merged and have the null termination discussion elsewhere.

Reply:
Sounds good to me.